Update bluetooth_le_tracker to use Bleak

bdraco opened the pull request:
Needs https://github.com/home-assistant/core/pull/75381

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update bluetooth_le_tracker to use Bleak

I wanted to keep the change as small as possible so its still a legacy tracker but using Bleak now.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Update bluetooth_le_tracker to use Bleak

bdraco commented:
We probably still need a periodic poll of `async_discovered_service_info` to account for devices that the bluez backend didn't send us updates for since they had no RSSI changes

MartinHjelmare commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -160,54 +131,62 @@ def discover_ble_devices():
+        mac: str,
+        now: datetime,
+        service_info: bluetooth.BluetoothServiceInfo,
+    ):
```
```suggestion
    ) -> None:
```

MartinHjelmare commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -160,54 +131,62 @@ def discover_ble_devices():
-        track_point_in_utc_time(hass, update_ble, dt_util.utcnow() + interval)
+    cancel = bluetooth.async_register_callback(hass, _async_update_ble, None)
+
+    def handle_stop(event: Event) -> None:
```
```suggestion
    @callback
    def handle_stop(event: Event) -> None:
```

MartinHjelmare commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -53,33 +51,23 @@
     """Set up the Bluetooth LE Scanner."""
 
     new_devices: dict[str, dict] = {}
     hass.data.setdefault(DATA_BLE, {DATA_BLE_ADAPTER: None})
```
```suggestion
```

bdraco committed:
Update homeassistant/components/bluetooth_le_tracker/device_tracker.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/bluetooth_le_tracker/device_tracker.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/bluetooth_le_tracker/device_tracker.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Update homeassistant/components/bluetooth_le_tracker/device_tracker.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Merge branch 'dev' into btle_tracker

bdraco commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -105,28 +93,13 @@ def see_device(address, name, new_device=False, battery=None):
-        return devices
-
     yaml_path = hass.config.path(YAML_DEVICES)
     devs_to_track = []
```
Good catch.  Those could be some long linear searches under the hood

bdraco committed:
set

bdraco committed:
still do interval updates to ensure everything stays home

bdraco committed:
still do interval updates to ensure everything stays home

bdraco committed:
still do interval updates to ensure everything stays home

bdraco committed:
naming

bdraco committed:
naming

bdraco committed:
adapt test

bdraco committed:
Add mock_bluetooth fixture

- Breaks this out from the bluetooth integration so
  other integrations can mock out bluetooth

- Blocks #75013

bdraco committed:
Merge branch 'bluetooth_fixtures' into btle_tracker

bdraco committed:
mock bluetooth in test

bdraco committed:
mock bluetooth in test

bdraco committed:
Merge branch 'dev' into btle_tracker

bdraco committed:
Merge branch 'dev' into btle_tracker

bdraco committed:
Merge branch 'dev' into btle_tracker

bdraco commented:
Looks like this had partial test coverage before.  Should add a few more when I don't have so many irons in the fire

bdraco committed:
Add device and advertisement to BluetoothServiceInfoBleak

bdraco committed:
convert address to bledevice

bdraco committed:
convert address to bledevice

bdraco commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -105,109 +103,107 @@ def see_device(address, name, new_device=False, battery=None):
-            battery = None
+        battery = None
+        try:
+            async with BleakClient(mac) as client:
```
This needs to use the BLEDevice instead if we can to avoid the client creating another scanner

Needs https://github.com/home-assistant/core/pull/75381

bdraco committed:
Merge branch 'service_info_with_bleak' into btle_tracker

bdraco committed:
fixes

bdraco committed:
cover

bdraco commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -105,109 +104,110 @@ def see_device(address, name, new_device=False, battery=None):
+            _LOGGER.warning(
+                "Timeout when trying to get battery status for %s", service_info.name
+            )
+        except (AttributeError, BleakError) as err:
```
```suggestion
        # Bleak currently has a few places where dbus attributes
        # can raise when there is another error. We need to trap them
        # until bleak releases v0.15+ which resolves these
        except (AttributeError, BleakError) as err:
```

bdraco committed:
Update homeassistant/components/bluetooth_le_tracker/device_tracker.py

bdraco commented on homeassistant/components/bluetooth_le_tracker/device_tracker.py:
```diff
@@ -139,15 +140,18 @@ async def _async_see_update_ble_battery(
                 "Timeout when trying to get battery status for %s", service_info.name
             )
-        except BleakError as err:
+        except (AttributeError, BleakError) as err:
```
Added a comment

MartinHjelmare commented:
This isn't ready for merge yet, right. The Bluetooth changes is from the other branch, I think.

bdraco committed:
Merge branch 'dev' into btle_tracker
